fix(start): fixed route scripts filtering

3fuyang opened the pull request:
The filtering before is like:

```ts
// ...
select: (state) => {
  // Assume here:
  // `[[undefined], undefined]`
  const scriptsOfEachMatch = state.matches.map((match) => match.scripts)
  // Filtered:
  // `[[undefined]]`
  const truthyScriptsOfEachMatch = scriptsOfEachMatch.filter(Boolean)
  // Turns out to be:
  // `[undefined]`
  // which will cause `TypeError` when `undefined` gets destructured below
  const truthyScripts = truthyScriptsOfEachMatch.flat(1)

  return {
    scripts: truthyScripts.map(({ children, ...script }) => ({ // <- TypeError: Cannot destructure property 'children' of 'undefined' as it is undefined.
      // ...
    })),
  }
},
// ...
```

Which will cause error when specifying `scripts` like:

```ts
export const Route = createRootRoute({
  head: () => ({
    // ...
    scripts: [undefined], // Passes type check but will panic at runtime
  }),
  // ...
})
```

This PR moves the `filter(Boolean)` after flattening. But I'm not sure if the following patterns are equivalent:

```js
arr.filter(Boolean).flat(1).filter(Boolean) // Semantically correct

arr.flat(1).filter(Boolean) // Less code
```

schiller-manuel commented:
we need an e2e test for this

3fuyang commented:
Hi @schiller-manuel, it seems the existing e2e are all related to examples, so I'm not sure where to put the e2e case for this. But I found the `start` package already includes unit tests for `scripts`, so I added it there instead. PTAL

schiller-manuel commented:
I would put an e2e test in this project: https://github.com/TanStack/router/tree/main/e2e/start/basic

add new routes for this setup and then access them via the playwright test and add some assertions

3fuyang commented:
Hi @schiller-manuel , thanks for the guidance!

I've added a new `/scripts` route and the corresponding case to the `start/basic` example. Here are a few tips for review:

- Two empty script files `script.js` and `script2.js` are added to `public` for testing.
- I have to include the added script files to `TSConfig.include` so that it does not conflict with the global ESLint configuration for typed linting.
- I found that [a case](https://github.com/TanStack/router/pull/3082/commits/523739ff5ee2a71badd1e3dd464caea55c44b414#diff-1be0197cd6a94732428eea7195281e42557356c49390484bc8a29126cdbcb0edL196) was marked with `only`, thus only this case gets evaluated when testing. I removed the `only` declaration and please let me know if this was intentional.
- The `invoking a server function with custom response status code` case fails when running e2e tests, so I modified the [assertion](https://github.com/TanStack/router/pull/3082/commits/a22446bf512adcdf70faf8c808b08acf3d8cabd9#diff-1be0197cd6a94732428eea7195281e42557356c49390484bc8a29126cdbcb0edL93) to make it pass.

BTW please let me know whether I should remove the UT added to the `start` package.

3fuyang commented:
Oops, the test seems to behave differently between headless and UI mode, ill check it

schiller-manuel commented:
very nice, thanks a lot!
